propulsion-hotel/Domain/GroupCheckout.fs

```diff
@@ -0,0 +1,115 @@
+
+    let initial = { pending = [||]; checkedOut = [||]; failed = [||]; payments = [||]; balance = 0m; completed = false }
+
+    let private removePending xs state = { state with pending = state.pending |> Array.except xs  }
+    let evolve state = function
+        | StaysAdded e ->
+            { state with pending = Array.append state.pending e.stays }
+        | StaysMerged e ->
+            { removePending (seq { for s in e.residuals -> s.stay }) state with
+                checkedOut = Array.append state.checkedOut e.residuals
+                balance = state.balance + (e.residuals |> Seq.sumBy (fun x -> x.residual)) }
+        | MergesFailed e ->
+            { removePending e.stays state with
+                failed = Array.append state.failed e.stays }
+        | Paid e ->
+            { state with
+                balance = state.balance - e.amount
+                payments = [| yield! state.payments; e.paymentId |] }
+        | Confirmed _ ->
+            { state with completed = true }
+    let fold : State -> Events.Event seq -> State = Seq.fold evolve
+
+module Flow =
+
+    type Action =
+        | MergeStays of           GuestStayId[]
+        | Ready of              balance : decimal
+        | Finished
+
+    let nextAction : Fold.State -> Action = function
+        | { completed = true } -> Finished
+        | { pending = xs } when not (Array.isEmpty xs) -> MergeStays xs
+        | { balance = bal } -> Ready bal
+        
+    let decide handleAction (state : Fold.State) : Async<'R * Events.Event list> =
+        nextAction state |> handleAction
+
+module Decide =
+
+    open Fold
+    
+    let add at stays state =
+        let registered = HashSet(seq { yield! state.pending; yield! state.failed; for x in state.checkedOut do yield x.stay })
+        match stays |> Array.except registered with
+        | [||] -> []
+        | xs -> [ Events.StaysAdded {| at = at; stays = xs |} ]
+
+    [<NoComparison; NoEquality>]
+    type ConfirmResult = Processing | Ok | BalanceOutstanding of decimal
+    let confirm at state =
+        match Flow.nextAction state with
+        | Flow.Finished -> ConfirmResult.Ok, []
+        | Flow.Ready 0m -> ConfirmResult.Ok, [ Events.Confirmed {| at = at |} ]
+        | Flow.Ready amount -> ConfirmResult.BalanceOutstanding amount, []
+        | Flow.MergeStays _ -> ConfirmResult.Processing, []
+
+    let pay paymentId amount at = function
+        | { payments = paymentIds } when paymentIds |> Array.contains paymentId -> []
+        | _ -> [ Events.Paid {| at = at; paymentId = paymentId; amount = amount |} ]
+
+type Service internal (resolve : GroupCheckoutId -> Equinox.Decider<Events.Event, Fold.State>) =
+
+    /// Called within Reactor host to Dispatch any relevant Reaction activities
+    member _.React(id, handleReaction) : Async<'R * int64> =
+        let decider = resolve id
+        decider.TransactExAsync((fun c -> Flow.decide handleReaction c.State), fun r c -> r, c.Version)
```

Comment:
Clarifying parens. It's unclear whether `c.Version` is a third argument to `TransactExAsync` at a glance without these parens

```suggestion
        decider.TransactExAsync((fun c -> Flow.decide handleReaction c.State), fun r c -> (r, c.Version))
```

Reply:
Good idea. Sometimes I add extension methods to Decider e.g. something like `TransactAsyncWithPostVersion` that make whats going on clearer aside from the sheer complexity of the expressions itself

Reply:
Added TranactAsyncWithPostVersion